src/react.tsx

```diff
@@ -0,0 +1,76 @@
+import React, { useCallback, useContext, useMemo, useRef } from "react";
+import { WTEventParams } from "./client";
+
+type ContextValue<Params = Record<string, any>> = {
+  params: React.RefObject<Partial<Params>>;
+  track: (overrides?: Partial<Params>) => void;
+};
+
+type Options<Params> = {
+  validateEvent?: (params: Partial<Params>) => boolean;
+};
+
+export const createProvider = <
+  Params extends Record<string | number, any> = WTEventParams
+>(
+  track: (params: Params) => void,
+  { validateEvent }: Options<Params> = {}
+) => {
+  const WTContext = React.createContext<ContextValue<Params>>({
+    params: { current: {} },
+    track: () => {
+      throw new Error("No WTProvider found");
+    },
+  });
+
+  const WTProvider: React.FC<{ params: Partial<Params> }> = ({
+    params,
+    children,
+  }) => {
+    const paramRef = useRef(params);
+    const { params: parentParams } = useContext(WTContext);
+
+    paramRef.current = { ...parentParams.current, ...params };
+
+    const localTrack = useCallback((overrides?: Partial<Params>) => {
+      const trackedParams = {
+        ...paramRef.current,
+        ...overrides,
+      };
+      if (!validateEvent || validateEvent(trackedParams)) {
+        track(trackedParams as Params);
+      }
+    }, []);
+
+    const contextValue = useMemo(
+      () => ({
+        track: localTrack,
+        params: paramRef,
+      }),
+      [localTrack]
+    );
+
+    return (
+      <WTContext.Provider value={contextValue}>{children}</WTContext.Provider>
+    );
+  };
+
+  const useTrack = (defaultParams: Partial<Params> = {}) => {
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    const { track } = useContext(WTContext);
+
+    return (params: Partial<Params>) => track({ ...defaultParams, ...params });
+  };
+
+  const useWTContext = () => {
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    const { track, params } = useContext(WTContext);
+
+    return {
+      track,
+      params: params.current,
+    };
+  };
+
+  return { useWT: useWTContext, WTProvider, useTrack };
```

Comment:
don't have a strong opinion on this I suppose, but curious why we don't call the function `useWT` and avoid the rename here?

Reply:
06bfb29d391079e84b8c275fdce7a301bec08bc5